### ISSUE TITLE
[Bug]: On Storybook 7, the code snippet isn't updated according with the current mapping selected value 

### ISSUE THREAD

**Issue by macabeus** [storybookjs/storybook#22100]
### Describe the bug

I have a React component whose one of the props receives a JSX.
Since we can't handle JSX directly on the control, I'm using the `mapping`, but I noticed that the code snippet doesn't output the JSX correctly.

It shows the mapping key instead of the current value and does not update when a new entry is picked.

The same code works on Storybook 6.

May be related to #22099

<img width="1037" alt="image" src="https://user-images.githubusercontent.com/9501115/232120877-53cbbd70-019a-4459-867d-db9f0feda221.png">

### To Reproduce

```jsx
import { Meta, StoryObj } from '@storybook/react';
import React from 'react';

const Example = (props) => {
  return <div>{props.buttons}</div>;
};

type Component = typeof Example;

const meta: Meta<Component> = {
  component: Example,
  argTypes: {
    buttons: {
      message: {
        control: 'text',
        if: { arg: 'withIcon' },
      },
      options: ['No Button', 'One Button', 'Two Buttons'],
      mapping: {
        'No Button': null,

        'One Button': (
          <>
            <button>Action 1</button>
          </>
        ),

        'Two Buttons': (
          <>
            <button>Action 1</button>
            <button>Action 2</button>
          </>
        ),
      },
      control: { type: 'select' },
    },
  },
};

export default meta;

export const Default: StoryObj<Component> = {
  args: {
    buttons: 'One Button',
  },
};
```

### System

```shell
Environment Info:

  System:
    OS: macOS 13.2
    CPU: (10) arm64 Apple M1 Max
  Binaries:
    Node: 16.15.1 - ~/.nvm/versions/node/v16.15.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.15.1/bin/yarn
    npm: 8.11.0 - ~/.nvm/versions/node/v16.15.1/bin/npm
  Browsers:
    Chrome: 111.0.5563.110
    Firefox: 109.0.1
    Safari: 16.3
  npmPackages:
    @storybook/addon-actions: ^7.0.4 => 7.0.4
    @storybook/addon-docs: ^7.0.4 => 7.0.4
    @storybook/addon-essentials: ^7.0.4 => 7.0.4
    @storybook/addon-links: ^7.0.4 => 7.0.4
    @storybook/addon-mdx-gfm: ^7.0.4 => 7.0.4
    @storybook/node-logger: ^7.0.4 => 7.0.4
    @storybook/react: ^7.0.4 => 7.0.4
    @storybook/react-vite: ^7.0.4 => 7.0.4
```

**Comment by shilman** [comment]
The source block has a `type` parameter: https://storybook.js.org/docs/react/api/doc-block-source#type

From the output it looks like you have set the type to `code`, meaning that it renders the source code of the story file. If you're using the default render function or a render function that accepts args, I'd expect to see JSX in the source block, and it would update dynamically as you modify the controls.

**Comment by macabeus** [comment]
Yes. I was expecting to see the JSX in the source block.

On my Storybook configuration, I didn't set any value for `type`. It was using the default one (`auto`).
I did a quick test setting `dynamic`, and it says that there is no code available:
<img width="1054" alt="image" src="https://user-images.githubusercontent.com/9501115/232217440-ac2d5f3e-db6c-41f2-9619-930b2a3b650d.png">
Is it expected?

**Comment by shilman** [comment]
Commented on your other issue. I think that's the underlying problem and this can probably be closed as a duplicate to that one.

**Comment by macabeus** [comment]
Closed since it seems to be duplicated with #22099
